aiida/tools/importexport/archive/common.py

```diff
@@ -61,3 +78,302 @@ def detect_archive_type(in_path: str) -> str:
+            # this stored in memory (no disk write)
+            folder.write_text('path/to/file.txt', 'content')
+            # this will be read from memory
+            text = folder.read_text('path/to/file.txt')
+
+        # all path/to/file.txt will now have been written to disk
+
+    """
+
+    def __init__(self, path: Union[Path, str], *, encoding: str = 'utf8'):
+        """Initialise cached folder.
+
+        :param path: folder path to cache
+        :param encoding: encoding of text to read/write
+
+        """
+        self._path = Path(path)
+        # dict mapping path -> (type, content)
+        self._cache = OrderedDict()  # type: ignore
+        self._encoding = encoding
+        self._max_items = 100  # maximum limit of files to store in memory
+
+    def _write_object(self, path: str, ctype: str, content: Any):
+        """Write an object from the cache to disk.
+
+        :param path: relative path of file
+        :param ctype: the type of the content
+        :param content: the content to write
+
+        """
+        if ctype == 'text':
+            text = content
+        elif ctype == 'json':
+            text = json.dumps(content)
+        else:
+            raise TypeError(f'Unknown content type: {ctype}')
+
+        (self._path / path).write_text(text, encoding=self._encoding)
+
+    def flush(self):
+        """Flush the cache."""
+        for path, (ctype, content) in self._cache.items():
+            self._write_object(path, ctype, content)
+
+    def _limit_cache(self):
+        """Ensure the cache does not exceed a set limit.
+
+        Content is uncached on a First-In-First-Out basis.
+
+        """
+        while len(self._cache) > self._max_items:
+            path, (ctype, content) = self._cache.popitem(last=False)
+            self._write_object(path, ctype, content)
+
+    def get_path(self, flush=True) -> Path:
+        """Return the path.
+
+        :param flush: flush the cache before returning
+
+        """
+        if flush:
+            self.flush()
+        return self._path
+
+    def write_text(self, path: str, content: str):
+        """write text to the cache.
+
+        :param path: path relative to base folder
+
+        """
+        assert isinstance(content, str)
+        self._cache[path] = ('text', content)
+        self._limit_cache()
+
+    def read_text(self, path) -> str:
+        """write text from the cache or base folder.
+
+        :param path: path relative to base folder
+
+        """
+        if path not in self._cache:
+            return (self._path / path).read_text(self._encoding)
+        ctype, content = self._cache[path]
+        if ctype == 'text':
+            return content
+        if ctype == 'json':
+            return json.dumps(content)
+
+        raise TypeError(f"content of type '{ctype}' could not be converted to text")
+
+    def write_json(self, path: str, data: dict):
+        """write dict to the cache as json.
+
```

Comment:
Add something like:

Dictionary is stored in memory until cache is flushed, at which point the dictionary is serialized to json and written to disk.

Reply:
added